core-test/src/test/scala/org/bitcoins/core/crypto/SchnorrTest.scala

```diff
@@ -0,0 +1,197 @@
+
+        // Assert our R value was the one used by comparing x-coordinates
+        assert(rBytes.tail == ByteVector(rKey))
+
+        // Compute the public key associated with s from public information
+        // (s*G = R + m*P)
+        val sPub: Array[Byte] =
+          NativeSecp256k1.computeSchnorrPubKey(messageArr,
+                                               rBytes.toArray,
+                                               publicKey.bytes.toArray)
+        val sPubKey: ECPublicKey = ECPublicKey.fromBytes(ByteVector(sPub))
+
+        // Make sure that NativeSecp256k1 doesn't deviate from Schnorr
+        assert(
+          sPubKey == Schnorr.computePubKey(message, nonce.publicKey, publicKey))
+
+        // Compute the public key associated with s from s itself (private info)
+        val realSPriv = ECPrivateKey.fromBytes(ByteVector(s))
+        val realSPub = realSPriv.publicKey
+
+        // Assert that both ways of computing s*G are equal
+        assert(sPubKey == realSPub)
+    }
+  }
+
+  /* Schnorr signatures have the property that if two messages are signed with the same keys,
+   * then those keys are leaked:
+   *
+   * sig1 = nonce + message1*privKey
+   * sig2 = nonce + message2*privKey
+   *
+   * => sig1 - sig2 = (message1 - message2)*privKey
+   * => privKey = (sig1 - sig2) * inverse(message1 - message2)
+   */
+  it should "leak keys if two messages are signed" in {
+    // The order of the secp256k1 curve (and thus the modulus of the field elements)
+    val M = BigInt(
+      "115792089237316195423570985008687907852837564279074904382605163141518161494337")
+
+    /** Returns num % M as a positive integer */
+    def modM(num: BigInt): BigInt = {
+      if (num < 0) {
+        (num % M) + M
+      } else {
+        num % M
+      }
+    }
+
+    /** Computes the inverse (mod M) of the input using the Euclidean Algorithm (log time)
+      * Cribbed from [[https://www.geeksforgeeks.org/multiplicative-inverse-under-modulo-m/]]
+      */
+    def modInverse(aInit: BigInt): BigInt = {
+      var a = modM(aInit)
+      var m = M
+      var m0 = m
+      var y = BigInt(0)
+      var x = BigInt(1)
+
+      if (m == 1)
+        return 0
+
+      while (a > 1) {
+        // q is quotient
+        val q = a / m
+
+        var t = m
+
+        // m is remainder now, process
+        // same as Euclid's algo
+        m = a % m
+        a = t
+        t = y
+
+        // Update x and y
+        y = x - q * y
+        x = t
+      }
+
+      if (x < 0) {
+        x += m0
+      }
+
+      modM(x)
+    }
+
+    forAll(CryptoGenerators.nonZeroPrivKey,
+           CryptoGenerators.schnorrNonce,
+           NumberGenerator.bytevector(32),
+           NumberGenerator.bytevector(32)) {
+      case (privKey, nonce, message1, message2) =>
+        // This will only work if we sign two different messages
+        assert(message1 != message2)
+
+        // Sign both messages using the same privKey and nonce
```

Comment:
the top level comment is wrong, it says it just needs to be the same private key. It needs to be the same key _and_ nonce, right?

Reply:
"same keys", so not wrong technically but the nonce terminology seems to complicate things :/ I'll make it more clear